In [4]:
import pandas as pd
import re
import string
import nltk

from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer

nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\ruchi\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\ruchi\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\ruchi\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [5]:
# Load dataset
df = pd.read_csv("Tweets-Copy1.csv")

In [6]:
df.head()

,tweet_id,airline_sentiment,airline_sentiment_confidence,negativereason,negativereason_confidence,airline,airline_sentiment_gold,name,negativereason_gold,retweet_count,text,tweet_coord,tweet_created,tweet_location,user_timezone
0,570306133677760513,neutral,1.0000,NaN,NaN,Virgin America,NaN,cairdin,NaN,0,@VirginAmerica What @dhepburn said.,NaN,2015-02-24 11:35:52 -0800,NaN,Eastern Time (US & Canada)
1,570301130888122368,positive,0.3486,NaN,0.0000,Virgin America,NaN,jnardino,NaN,0,@VirginAmerica plus you've added commercials t...,NaN,2015-02-24 11:15:59 -0800,NaN,Pacific Time (US & Canada)
2,570301083672813571,neutral,0.6837,NaN,NaN,Virgin America,NaN,yvonnalynn,NaN,0,@VirginAmerica I didn't today... Must mean I n...,NaN,2015-02-24 11:15:48 -0800,Lets Play,Central Time (US & Canada)
3,570301031407624196,negative,1.0000,Bad Flight,0.7033,Virgin America,NaN,jnardino,NaN,0,@VirginAmerica it's really aggressive to blast...,NaN,2015-02-24 11:15:36 -0800,NaN,Pacific Time (US & Canada)
4,570300817074462722,negative,1.0000,Can't Tell,1.0000,Virgin America,NaN,jnardino,NaN,0,@VirginAmerica and it's a really big bad thing...,NaN,2015-02-24 11:14:45 -0800,NaN,Pacific Time (US & Canada)


In [7]:
df.columns

Index(['tweet_id', 'airline_sentiment', 'airline_sentiment_confidence',
       'negativereason', 'negativereason_confidence', 'airline',
       'airline_sentiment_gold', 'name', 'negativereason_gold',
       'retweet_count', 'text', 'tweet_coord', 'tweet_created',
       'tweet_location', 'user_timezone'],
      dtype='object')

In [8]:
df.isnull().sum()


tweet_id                            0
airline_sentiment                   0
airline_sentiment_confidence        0
negativereason                   5462
negativereason_confidence        4118
airline                             0
airline_sentiment_gold          14600
name                                0
negativereason_gold             14608
retweet_count                       0
text                                0
tweet_coord                     13621
tweet_created                       0
tweet_location                   4733
user_timezone                    4820
dtype: int64

In [9]:
# exploring no.of samples nd columns
print("Shape:", df.shape) 
print("\nColumns:", df.columns)
print("\nInfo:")
df.info()

Shape: (14640, 15)

Columns: Index(['tweet_id', 'airline_sentiment', 'airline_sentiment_confidence',
       'negativereason', 'negativereason_confidence', 'airline',
       'airline_sentiment_gold', 'name', 'negativereason_gold',
       'retweet_count', 'text', 'tweet_coord', 'tweet_created',
       'tweet_location', 'user_timezone'],
      dtype='object')

Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14640 entries, 0 to 14639
Data columns (total 15 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   tweet_id                      14640 non-null  int64  
 1   airline_sentiment             14640 non-null  object 
 2   airline_sentiment_confidence  14640 non-null  float64
 3   negativereason                9178 non-null   object 
 4   negativereason_confidence     10522 non-null  float64
 5   airline                       14640 non-null  object 
 6   airline_sentiment_gold        40 non-null     ob

In [10]:

 # class distribution
df['airline_sentiment'].value_counts()



airline_sentiment
negative    9178
neutral     3099
positive    2363
Name: count, dtype: int64

In [11]:
# sample texts
print("\nSample Texts:")
print(df['text'].head(10))


Sample Texts:
0                  @VirginAmerica What @dhepburn said.
1    @VirginAmerica plus you've added commercials t...
2    @VirginAmerica I didn't today... Must mean I n...
3    @VirginAmerica it's really aggressive to blast...
4    @VirginAmerica and it's a really big bad thing...
5    @VirginAmerica seriously would pay $30 a fligh...
6    @VirginAmerica yes, nearly every time I fly VX...
7    @VirginAmerica Really missed a prime opportuni...
8      @virginamerica Well, I didn't…but NOW I DO! :-D
9    @VirginAmerica it was amazing, and arrived an ...
Name: text, dtype: object


In [12]:
# NLP Preprocessing
import sys
!{sys.executable} -m pip install nltk


[notice] A new release of pip is available: 25.0.1 -> 26.0.1
[notice] To update, run: C:\Users\ruchi\AppData\Local\Programs\Python\Python312\python.exe -m pip install --upgrade pip


In [13]:
import nltk
from nltk.corpus import stopwords
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\ruchi\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [27]:
from nltk.corpus import stopwords

stop_words = set(stopwords.words('english'))
lemmatizer = WordNetLemmatizer()

In [28]:
# Lowercasing
def to_lower(text):
    return text.lower()
    

In [29]:
# Remove URLs & Special Characters
def remove_urls(text):
    return re.sub(r'http\S+|www\S+|https\S+', '', text)

def remove_special_chars(text):
    return re.sub(r'[^a-zA-Z\s]', '', text)

In [32]:
#removing punctuation
def remove_punctuation(text):
    return text.translate(str.maketrans('', '', string.punctuation))


In [33]:
# Tokenization
def tokenize(text):
    return word_tokenize(text)

In [19]:
# Stemming
def remove_stopwords(tokens):
    return [word for word in tokens if word not in stop_words]

In [20]:
df['text'] = df['text'].fillna("")

In [21]:
from nltk.stem import WordNetLemmatizer

lemmatizer = WordNetLemmatizer()

def lemmatize(tokens):
    return [lemmatizer.lemmatize(word) for word in tokens]

In [34]:
def preprocess(text):
    if not isinstance(text, str):
        return ""
    
    text = to_lower(text)
    text = remove_urls(text)
    text = remove_punctuation(text)
    text = remove_special_chars(text)
    
    tokens = tokenize(text)
    tokens = remove_stopwords(tokens)
    tokens = lemmatize(tokens)
    
    return " ".join(tokens)

In [35]:

import nltk
nltk.download('punkt_tab')

[nltk_data] Downloading package punkt_tab to
[nltk_data]     C:\Users\ruchi\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


True

In [36]:
df['clean_text'] = df['text'].apply(preprocess)

In [37]:
import sys
print(sys.executable)

C:\Users\ruchi\AppData\Local\Programs\Python\Python312\python.exe


In [38]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [39]:
# Feature Engineering (TF-IDF)
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer(max_features=5000)

X = tfidf.fit_transform(df['clean_text'])
y = df['airline_sentiment']

In [40]:
# Train-Test Split
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [54]:
# Model Building
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import MultinomialNB
from sklearn.tree import DecisionTreeClassifier

# Logistic Regression
lr = LogisticRegression(max_iter=200)
lr.fit(X_train, y_train)

# Naive Bayes
nb = MultinomialNB()
nb.fit(X_train, y_train)

# Decision Tree
dt = DecisionTreeClassifier()
dt.fit(X_train, y_train)

# train model
dt = DecisionTreeClassifier()
dt.fit(X_train, y_train)

# evaluate 
evaluate(dt, "Decision Tree")

--- Decision Tree ---
Accuracy: 0.6960382513661202
Precision: 0.6921088891955888
Recall: 0.6960382513661202
F1 Score: 0.6939459837378924



In [44]:
# 6. Model Evaluation
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

def evaluate(model, name):
    y_pred = model.predict(X_test)
    
    print(f"--- {name} ---")
    print("Accuracy:", accuracy_score(y_test, y_pred))
    print("Precision:", precision_score(y_test, y_pred, average='weighted'))
    print("Recall:", recall_score(y_test, y_pred, average='weighted'))
    print("F1 Score:", f1_score(y_test, y_pred, average='weighted'))
    print()

evaluate(lr, "Logistic Regression")
evaluate(nb, "Naive Bayes")
evaluate(dt, "Decision Tree")

--- Logistic Regression ---
Accuracy: 0.8012295081967213
Precision: 0.7920669047557122
Recall: 0.8012295081967213
F1 Score: 0.7891284165113395

--- Naive Bayes ---
Accuracy: 0.7387295081967213
Precision: 0.7502507518447963
Recall: 0.7387295081967213
F1 Score: 0.6930483167803582

--- Decision Tree ---
Accuracy: 0.6953551912568307
Precision: 0.6920823312061359
Recall: 0.6953551912568307
F1 Score: 0.6936110826488899



In [59]:
df['predicted'] = lr.predict(X)

print(df[['airline', 'text', 'airline_sentiment', 'predicted']].head())

          airline                                               text  \
0  Virgin America                @VirginAmerica What @dhepburn said.   
1  Virgin America  @VirginAmerica plus you've added commercials t...   
2  Virgin America  @VirginAmerica I didn't today... Must mean I n...   
3  Virgin America  @VirginAmerica it's really aggressive to blast...   
4  Virgin America  @VirginAmerica and it's a really big bad thing...   

  airline_sentiment predicted  
0           neutral   neutral  
1          positive  positive  
2           neutral  negative  
3          negative  negative  
4          negative  negative  


In [64]:
# overall Accuracy testing
from sklearn.metrics import accuracy_score

print("Accuracy:", accuracy_score(df['airline_sentiment'], df['predicted']))

Accuracy: 0.8485655737704918


 Insights & Analysis

Decision Tree Performance
- Moderate accuracy (~69.6%) on test data.  
- Precision and recall are roughly equal, meaning the model is balanced across predicting positive, negative, and neutral sentiments.  
- F1 score (~0.694) confirms that overall performance is consistent but not perfect.  

 Data Imbalance Effect
- Airline sentiment datasets often have more neutral/negative than positive tweets, which can influence metrics.  
- The higher accuracy on the full dataset (~0.848) may be biased toward dominant classes.  

 Feature Representation
- TF-IDF captures the most frequent terms, but contextual meaning is lost, so the model struggles on subtle sentiment expressions.  
- Using **Word2Vec**, **BERT embeddings**, or **n-grams** might improve performance.  

 Decision Tree Limitations
- Trees can overfit on training data.  
- Might benefit from **Random Forest** or **Gradient Boosting** for better generalization.  

Model Overall Comparison
- **Logistic Regression** performed best overall because it handles text features effectively.  
- **Naive Bayes** is fast but assumes independence between words, which is not always realistic.  
- **Decision Tree** gives moderate accuracy but tends to overfit on text data.  
- **TF-IDF** gave better results than Bag of Words as it considers word importance.  
  